File: devvit/src/server/maczine/urls.ts
/**
 * URL handling, so the same article compares equal however it was written.
 *
 * Pure functions: no Devvit runtime, so they can be unit tested directly.
 */

export const ALLOWED_HOSTS = ['mactechsolutionsllc.com', 'www.mactechsolutionsllc.com'];
export const ARTICLE_PATH_PREFIX = '/maczine/';

const TRACKING_EXACT = new Set([
  'fbclid',
  'gclid',
  'dclid',
  'msclkid',
  'ref',
  'ref_src',
  'igshid',
  'yclid',
]);
const TRACKING_PREFIXES = ['utm_', 'mc_'];

const isTracking = (key: string): boolean => {
  const k = key.toLowerCase();
  return TRACKING_EXACT.has(k) || TRACKING_PREFIXES.some((p) => k.startsWith(p));
};

const parse = (url: string): URL | undefined => {
  try {
    return new URL(url.trim());
  } catch {
    return undefined;
  }
};

const cleanQuery = (url: URL): string => {
  const kept = [...url.searchParams.entries()].filter(([k]) => !isTracking(k));
  kept.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  const params = new URLSearchParams(kept);
  const query = params.toString();
  return query ? `?${query}` : '';
};

/**
 * The comparison key: https, bare lowercase host, no trailing slash, no
 * fragment, no tracking parameters, remaining parameters sorted. Returns ''
 * for anything unparseable, which never equals a real article.
 */
export const normalizeUrl = (url: string): string => {
  const parsed = parse(url);
  if (!parsed) return '';
  const host = parsed.hostname.toLowerCase().replace(/^www\./, '');
  const port = parsed.port && parsed.port !== '80' && parsed.port !== '443' ? `:${parsed.port}` : '';
  const path = parsed.pathname.replace(/\/+$/, '');
  return `https://${host}${port}${path}${cleanQuery(parsed)}`;
};

/**
 * The URL actually submitted: the feed link as published, forced to https,
 * without tracking parameters, fragment or trailing slash. The host is kept
 * as-is so the post points at the exact live address.
 */
export const canonicalUrl = (url: string): string => {
  const parsed = parse(url);
  if (!parsed) return '';
  const path = parsed.pathname.replace(/\/+$/, '');
  return `https://${parsed.host.toLowerCase()}${path}${cleanQuery(parsed)}`;
};

/** True only for an article page on MacTech's own site. */
export const isMacZineArticle = (url: string): boolean => {
  const parsed = parse(url);
  if (!parsed) return false;
  if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') return false;
  if (!ALLOWED_HOSTS.includes(parsed.hostname.toLowerCase())) return false;
  if (parsed.port && parsed.port !== '80' && parsed.port !== '443') return false;
  if (!parsed.pathname.startsWith(ARTICLE_PATH_PREFIX)) return false;
  const slug = parsed.pathname.slice(ARTICLE_PATH_PREFIX.length).replace(/^\/+|\/+$/g, '');
  return slug.length > 0 && slug !== 'feed.xml';
};

/** Do two URLs point at the same article? */
export const sameArticle = (a: string, b: string): boolean => {
  const left = normalizeUrl(a);
  return left !== '' && left === normalizeUrl(b);
};

File: reddit/src/maczine_reddit/urls.py
"""URL normalization, so the same article compares equal however it was written."""

from __future__ import annotations

from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

ALLOWED_HOSTS = {"mactechsolutionsllc.com", "www.mactechsolutionsllc.com"}
ARTICLE_PATH_PREFIX = "/maczine/"

_TRACKING_EXACT = {"fbclid", "gclid", "dclid", "msclkid", "ref", "ref_src", "igshid", "yclid"}
_TRACKING_PREFIXES = ("utm_", "mc_")


def _is_tracking(key: str) -> bool:
    k = key.lower()
    return k in _TRACKING_EXACT or k.startswith(_TRACKING_PREFIXES)


def normalize_url(url: str) -> str:
    """Comparison key: https, bare lowercase host, no trailing slash, fragment
    or tracking parameters; remaining query parameters sorted."""
    parts = urlsplit(url.strip())
    host = (parts.hostname or "").lower().removeprefix("www.")
    if parts.port and parts.port not in (80, 443):
        host = f"{host}:{parts.port}"
    path = parts.path.rstrip("/") or ""
    query = sorted((k, v) for k, v in parse_qsl(parts.query, keep_blank_values=True) if not _is_tracking(k))
    return urlunsplit(("https", host, path, urlencode(query), ""))


def canonical_url(url: str) -> str:
    """The URL to submit: the feed link as published, forced to https, with
    tracking parameters, fragment and trailing slash removed. The host is
    kept as-is so the post points at the exact live address."""
    parts = urlsplit(url.strip())
    query = [(k, v) for k, v in parse_qsl(parts.query, keep_blank_values=True) if not _is_tracking(k)]
    return urlunsplit(("https", parts.netloc.lower(), parts.path.rstrip("/"), urlencode(query), ""))


def is_maczine_article(url: str) -> bool:
    """True only for an article page on MacTech's own site."""
    parts = urlsplit(url.strip())
    if parts.scheme not in ("http", "https"):
        return False
    if (parts.hostname or "").lower() not in ALLOWED_HOSTS or parts.port not in (None, 80, 443):
        return False
    slug = parts.path.removeprefix(ARTICLE_PATH_PREFIX).strip("/")
    return parts.path.startswith(ARTICLE_PATH_PREFIX) and bool(slug) and slug != "feed.xml"
